include/circt/Dialect/OM/OMOpInterfaces.td

```diff
@@ -0,0 +1,57 @@
+//===- OMOpInterfaces.td - Object Model dialect op interfaces -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This contains the Object Model dialect operation interfaces.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_OM_OMOPINTERFACES_TD
+#define CIRCT_DIALECT_OM_OMOPINTERFACES_TD
+
+include "mlir/IR/OpBase.td"
+
+def ClassLike : OpInterface<"ClassLike"> {
+  let cppNamespace = "circt::om";
+
+  let description = [{
+    Common functionality for class-like operations.
+  }];
+
+  let methods = [
+    InterfaceMethod<"Get the class-like symbol name",
+      "llvm::StringRef", "getSymName", (ins)>,
+    InterfaceMethod<"Get the class-like symbol name attribute",
+      "mlir::StringAttr", "getSymNameAttr", (ins)>,
+    InterfaceMethod<"Get the class-like symbol name attribute name",
+      "mlir::StringAttr", "getSymNameAttrName", (ins)>,
+    InterfaceMethod<"Get the class-like formal parameter names attribute",
+      "mlir::ArrayAttr", "getFormalParamNames", (ins)>,
+    InterfaceMethod<"Get the class-like formal parameter names attribute name",
+      "mlir::StringAttr", "getFormalParamNamesAttrName", (ins)>,
+    InterfaceMethod<"Get the class-like body region",
+      "mlir::Region &", "getBody", (ins)>,
+    InterfaceMethod<"Get the class-like body block",
+      "mlir::Block *", "getBodyBlockPointer", (ins),
```

Comment:
I wanted to define this as `getBodyBlock`, but the way ODS works wouldn't allow it. You can declare interface methods that are just implemented by ODS generated methods, but I couldn't see a way to declare an interface method that was implemented by an extraClassDeclaration on the op. If I move `getBodyBlock` from an extraClassDeclaration to an interface method, anywhere you directly create a ClassOp, you will need to cast it to a ClassLike in order to call this helper, which is kind of inconvenient. So my solution was to keep the extraClassDeclaration for convenience, but also add a convenience method to the interface with a different name, which is useful for the generic functions that work with the interface. Curious if there is a better way, or a better name.

Reply:
Maybe I'm missing something but can't we just define `getBodyBlock` to both `ClassLike` and `ClassOp`? Interface and class declaration should use different name space so I guess it compiles. `firrtl::FModuleLike::getParameters()` and `firrtl::FModuleOp::getParameters` are doing same thing.

Reply:
I think the one after `methodBody`, `defaultImplementation` is where you can put implementation that goes on the trait (=> accessible on the op)?

Reply:
Oh, I see, I was trying with `methodBody`, let me try with `defaultImplementation`, that would be great.

Reply:
It looks like I'm able to define `getBodyBlock` on the interface with `methodBody`, just like how `getParameters` works for `FModuleLike` and `FModuleOp`. Thanks for the pointers.